### ISSUE TITLE
Handling of binary files

### ISSUE THREAD

**Issue by kinoute** [scrapy-plugins/scrapy-playwright#184]
Hello,

While trying to upgrade my "normal" Scrapy CrawlSpider to use Playwright through this plug-in, I came across the handling of binary files. For the classic Scrapy, every file was processed the same way: the Scrapy response body was the bytes of the requested file – HTML page, PDF, Office document etc.

I understand that with Playwright, as it is literally a real browser, every "unreadable" document by the browser should trigger the `download` event – except PDFs which are read with the built-in viewer, but this is a [known issue](https://github.com/microsoft/playwright/issues/7822).

My concerns are more about how to handle this: with the plug-in, even if I try to catch a Playwright exception or play with the event handler `download`, the plug-in still raises an error and is unable to trigger the callback of the spider, like `parse_page`. Therefore, these "binary" files are never processed.

```
 Browser chromium launched
 Browser context started: 'custom' (persistent=False)
 New page created, page count is 1 (1 for all contexts)
 Request: <GET https://dornsife.usc.edu/assets/sites/298/docs/ir211wk12samplexls> (resource type: document, referrer: https://dornsife.usc.edu/)
 ***Caught exception: <class 'playwright_impl_api_typesError'>***
 Response: <200 https://dornsife.usc.edu/assets/sites/298/docs/ir211wk12samplexls> (referrer: None)
 ***Document triggered a download***
 <twistedpythonfailureFailure playwright_impl_api_typesError: net::ERR_ABORTED at https://dornsife.usc.edu/assets/sites/298/docs/ir211wk12samplexls
=========================== logs ===========================
navigating to "https://dornsife.usc.edu/assets/sites/298/docs/ir211wk12samplexls", waiting until "load"
============================================================>
 Closing spider (finished)
 Spider closed (finished)
 Closing download handler
 Closing download handler
Browser context closed: 'custom' (persistent=False)
```

Through a dedicated middleware for Playwright, here is what I tried:

**Add a download event handler**

```python3
   [...]
    async def handle_download(self, download):
        logger.info('Document triggered a download')
        await download.save_as(download.suggested_filename)
        self.binary_files[download.url] = download.suggested_filename

    async def process_request(
        self,
        request,
        spider,
    ) -> None:
        request.meta['playwright'] = True
        request.meta['playwright_include_page'] = True

        request.meta['playwright_page_event_handlers'] = {
            'download': self.handle_download,
        }

        return None

    async def process_exception(
        self,
        request,
        spider,
        exception,
    ):
        logger.info('Caught exception: %s', exception.__class__)
        if request.url in self.binary_files:
            logger.info('File found locally for this URL')


```

My idea was to:

1. Download the file locally with the `download` Playwright event handler
2. Store in a middleware attribute the URL as key and the filename as value to use it in the `process_exception` 
3. Return a custom Scrapy `Response` object to get the file processed as normal in `process_exception`

But as we can see in the log above, the `download` event is triggered _after_ the `process_exception` method, therefore, this is not possible.

I don't think I can add a custom Scrapy response in the `download` event handler too?

**Detect and redirect binary files to a classic Scrapy**

I figured I could:

* Detect the extension of a binary file in `process_request` and redirect it to a non-Playwright version of my spider ;
* Detect the mime type of a binary file in `process_response` – sometimes, files don't have extension –, and do the same.

Like so:

```python3

    def _is_request_for_binary_file(self, url: str) -> bool:
        lowercase_path = urlparse(url).path.lower()
        return any(lowercase_path.endswith(ext) for ext in self.binary_extensions)

    async def process_request(
        self,
        request,
        spider,
    ) -> None:
        if self.is_request_for_binary_file(request.url):
             logger.warn('[Playwright] Binary file requested')
             logger.warn('[Playwright] Transferring to normal Scrapy for crawling')
            # send crawling task to another scrapy instance here
             raise IgnoreRequest()

    async def process_response(
        self,
        request,
        response,
        spider,
    ):
        mimetype, *_ = response.headers['content-type'].decode().split(';')

        if mimetype in self.binary_mimetypes:
            logger.warn('Binary file detected')
            logger.warn('Sending file to normal Scrapy for crawling')
            # send crawling task to another scrapy instance here
            raise IgnoreRequest()

        return response
```

But I'm pretty sure there must be a better way. Any idea where we could avoid these exceptions and transform binary files as valid Scrapy responses, which will trigger the Request callback?

I could not reproduce it but in my early tests, the spider's traceback when trying to process a binary file was targeting the `download_request` method in the `handler.py` file of the plug-in.

Thank you very much in advance for any tips!

**Comment by elacuesta** [comment]
This is an interesting problem. I think it might be possible to handle it with [`Page.expect_download`](https://playwright.dev/python/docs/api/class-page#page-wait-for-download), something like this (the try/except is necessary for some reason, see [this SO answer](https://stackoverflow.com/a/74144570)):
```python
async with page.expect_download() as download_info:
    try:
        response = await page.goto(url=request.url, **page_goto_kwargs)
    except Exception:
        pass
download = await download_info.value
with tempfile.NamedTemporaryFile() as temp_file:
    await download.save_as(temp_file.name)
    binary_file_contents = temp_file.seek(0)
```

There are still (at least) two problems with this:
1. `Page.expect_download` just times out for "regular" requests, i.e. requests that just receive some HTML and do not initiate any download. We would need to mark these requests for which we expect a binary response, maybe with a `playwright_expect_download` meta key or something similar. I don't quite like that, I wish it would be possible to auto-detect this.
2. Since the `Page.goto` call fails, there is no response, hence there are no response headers :shrug:

**Comment by elacuesta** [comment]
@kinoute are you still interested in this? Would you be able to test the branch from #228? There should be no updates necessary in your code, and there's is a sample spider at https://github.com/scrapy-plugins/scrapy-playwright/blob/12e743de3f90bc73a13f475c2bdabeeedd1d6a74/examples/download.py.

**Comment by kinoute** [comment]
@elacuesta I'm still very interested, thank you for working on this! I will check your branch tomorrow and give you some feedbacks ASAP

**Comment by kinoute** [comment]
Sorry for the late answer!

I tried with [a PDF](https://defret.in/assets/certificates/attestation_secnumacademie.pdf) and I get this as a `response.body` :

```html
[Playwright] b'
<!DOCTYPE html>
<html>
<head></head>
<body style="height: 100%; width: 100%; overflow: hidden; margin:0px; background-color: rgb(82, 86, 89);">
<embed name="493517358C2EA724599C13DA0B827E42" style="position:absolute; left: 0; top: 0;" width="100%" height="100%" src="about:blank" type="application/pdf" internalid="493517358C2EA724599C13DA0B827E42">
</body>
</html>'
```

It might be linked to Chromium PDF built-in viewer that shows the PDF instead of downloading it?

I tried with [a PowerPoint file](https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt) also, got a different output (no body but status code 200):

```
[2023-09-10 07:08:27,707][DEBUG] [Context=custom] New page created, page count is 2 (2 for all contexts)
[2023-09-10 07:08:27,836][DEBUG] [Context=custom] Request: <GET https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt> (resource type: document, referrer: https://defret.in/)
[2023-09-10 07:08:28,488][DEBUG] [Context=custom] Response: <200 https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt>
[2023-09-10 07:08:29,204][WARNING] Navigating to <GET https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt> returned None, the response will have empty headers and status 200
[2023-09-10 07:08:29,317][ERROR] <twisted.python.failure.Failure builtins.KeyError: b'Content-Type'>
```

**Comment by elacuesta** [comment]
I'm sorry, I'm getting the response bodies as bytes. Can you share the full spider code and point to the commit you're using? For reference, I'm getting the following results with 57a8192bb76029508f1eb04e3e39be187c882f20 (the tip of the `handle-downloads` branch) by adding your pdf/ppt URLs in the [download](https://github.com/scrapy-plugins/scrapy-playwright/blob/57a8192bb76029508f1eb04e3e39be187c882f20/examples/download.py) sample spider:

```
2023-09-12 20:11:07 [scrapy.core.engine] INFO: Spider opened
2023-09-12 20:11:07 [scrapy.extensions.logstats] INFO: Crawled 0 pages (at 0 pages/min), scraped 0 items (at 0 items/min)
2023-09-12 20:11:07 [scrapy.extensions.telnet] INFO: Telnet console listening on 127.0.0.1:6023
2023-09-12 20:11:07 [scrapy-playwright] INFO: Starting download handler
2023-09-12 20:11:12 [scrapy-playwright] INFO: Launching browser chromium
2023-09-12 20:11:12 [scrapy-playwright] INFO: Browser chromium launched
2023-09-12 20:11:12 [scrapy-playwright] DEBUG: Browser context started: 'default' (persistent=False, remote=False)
2023-09-12 20:11:12 [scrapy-playwright] DEBUG: [Context=default] New page created, page count is 1 (1 for all contexts)
2023-09-12 20:11:12 [scrapy-playwright] DEBUG: [Context=default] New page created, page count is 2 (2 for all contexts)
2023-09-12 20:11:12 [scrapy-playwright] DEBUG: [Context=default] Request: <GET https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt> (resource type: document)
2023-09-12 20:11:12 [scrapy-playwright] DEBUG: [Context=default] Request: <GET https://defret.in/assets/certificates/attestation_secnumacademie.pdf> (resource type: document)
2023-09-12 20:11:12 [scrapy-playwright] DEBUG: [Context=default] Response: <200 https://defret.in/assets/certificates/attestation_secnumacademie.pdf>
2023-09-12 20:11:12 [scrapy-playwright] WARNING: Navigating to <GET https://defret.in/assets/certificates/attestation_secnumacademie.pdf> returned None, the response will have empty headers and status 200
2023-09-12 20:11:12 [scrapy.core.engine] DEBUG: Crawled (200) <GET https://defret.in/assets/certificates/attestation_secnumacademie.pdf> (referer: None) ['playwright']
2023-09-12 20:11:13 [scrapy.core.scraper] DEBUG: Scraped from <200 https://defret.in/assets/certificates/attestation_secnumacademie.pdf>
{'url': 'https://defret.in/assets/certificates/attestation_secnumacademie.pdf', 'response_cls': 'Response', 'first_bytes': b'%PDF-1.3\n%\xe2\xe3\xcf\xd3\n9 0 obj\n<< /Type /Page /Parent 1 0 R /LastMod', 'filename': 'attestation_secnumacademie.pdf'}
2023-09-12 20:11:13 [scrapy-playwright] DEBUG: [Context=default] Response: <200 https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt>
2023-09-12 20:11:13 [scrapy-playwright] WARNING: Navigating to <GET https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt> returned None, the response will have empty headers and status 200
2023-09-12 20:11:13 [scrapy.core.engine] DEBUG: Crawled (200) <GET https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt> (referer: None) ['playwright']
2023-09-12 20:11:13 [scrapy.core.scraper] DEBUG: Scraped from <200 https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt>
{'url': 'https://www.cdse.edu/Portals/124/Documents/jobaids/industrial/foreign-travel-brief-template.ppt', 'response_cls': 'Response', 'first_bytes': b'\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00>\x00\x03\x00\xfe\xff\t\x00\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x14\x00\x00\x00\xbd\t\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00', 'filename': 'foreign-travel-brief-template.ppt'}
2023-09-12 20:11:13 [scrapy.core.engine] INFO: Closing spider (finished)
```

**Comment by tommylge** [comment]
Hello @elacuesta it is working fine, thanks for the fix.
However we've experienced something strange regarding the response headers. 
Indeed, we need to know the mimetype from content-type, but we got an almost empty response headers, we can only access the date one. Have you anything in mind that could explain / fix this please. 

Thanks again for you fix.

**Comment by elacuesta** [comment]
Makes sense, the problem is the `Page.goto` call ([source](https://github.com/scrapy-plugins/scrapy-playwright/blob/57a8192bb76029508f1eb04e3e39be187c882f20/scrapy_playwright/handler.py#L447)) fails for downloads, so there is no response from which to get headers.

**Comment by kinoute** [comment]
Ok so I guess we will have to get the mime type by ourself when it is not provided, which is fine. Do you want us to make more tests or it's okay and you can merge this feature?

Thanks!

**Comment by tommylge** [comment]
Hello @elacuesta, do you have any eta about when this fix will be merged into main ?

**Comment by elacuesta** [comment]
Released as [v0.0.33](https://github.com/scrapy-plugins/scrapy-playwright/releases/tag/v0.0.33)
